### PR TITLE
[TASK] Use method findFirstPreviousOfNode instead of findFirstPrevious

### PR DESCRIPTION
**Pull request by sabbelasichon**
Relates: https://github.com/rectorphp/rector/issues/6441

### CODE REVIEW

**Comment by sabbelasichon** [comment]
@TomasVotruba If you think this is  the right change, i am gonna remove the WIP and we could fix it.

**Comment by TomasVotruba** [comment]
The title doesn't explain much what is going on. Could you elaborate what is this for?

**Comment by sabbelasichon** [comment]
Oh. I thought the body link says it all: rectorphp/rector#6441

It is about this issue here: https://getrector.org/demo/abf577a0-a6ca-4f8b-b394-f50c34558faa

I fixed it just by using this method call here https://github.com/rectorphp/rector-src/pull/86/files#diff-fe54c83c0449653c57f79459ef63e2e4c847e1bad68461add56de99d7db89e54R142 

But you are right, the title itself is not clear enough.

File: rules-tests/Renaming/Rector/Name/RenameClassRector/FixtureAutoImportNames/rename_class_with_same_name_but_different_namespace_multiple_use_statements.php.inc [hunk 1]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace Rector\Tests\Renaming\Rector\Name\RenameClassRector\FixtureAutoImportNames;
+
+use Rector\Core\Rector\AbstractRector;
+use Rector\Tests\Renaming\Rector\Name\RenameClassRector\Source\FirstNamespace\SomeServiceClass;
+use Doctrine\Instantiator\Instantiator;
```
Comment (TomasVotruba) [inline rules-tests/Renaming/Rector/Name/RenameClassRector/FixtureAutoImportNames/rename_class_with_same_name_but_different_namespace_multiple_use_statements.php.inc]: What is this use statement for?

File: rules/Renaming/NodeManipulator/ClassRenamer.php [hunk 2]
```diff
@@ -139,7 +139,7 @@ private function refactorName(Name $name, array $oldToNewClasses): ?Name
 
     private function removeUseName(Name $oldName): void
     {
-        $uses = $this->betterNodeFinder->findFirstPreviousOfNode(
+        $uses = $this->betterNodeFinder->findFirstPrevious(
```
Comment (TomasVotruba) [inline rules/Renaming/NodeManipulator/ClassRenamer.php:142]: It seems these 2 methods are doing the same thing. Maybe just one should exist.

What is the difference of these methods?

**Comment by TomasVotruba** [comment]
@sabbelasichon It's better to have the content in the PR instead of stream of links, just so people who have no idea (like me) can understand what is going on without clinking them.

Usually the better the description and title is, the faster merge is :), as the full context is here and all questions are answered already.

So this PR is removing the unused import if class is renamed?

File: rules-tests/Renaming/Rector/Name/RenameClassRector/FixtureAutoImportNames/rename_class_with_same_name_but_different_namespace_multiple_use_statements.php.inc [hunk 3]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace Rector\Tests\Renaming\Rector\Name\RenameClassRector\FixtureAutoImportNames;
+
+use Rector\Core\Rector\AbstractRector;
```
Comment (TomasVotruba) [inline rules-tests/Renaming/Rector/Name/RenameClassRector/FixtureAutoImportNames/rename_class_with_same_name_but_different_namespace_multiple_use_statements.php.inc]: What is this use for?

File: rules-tests/Renaming/Rector/Name/RenameClassRector/FixtureAutoImportNames/rename_class_with_same_name_but_different_namespace_multiple_use_statements.php.inc (hunk 1 above)
Comment (sabbelasichon) [reply rules-tests/Renaming/Rector/Name/RenameClassRector/FixtureAutoImportNames/rename_class_with_same_name_but_different_namespace_multiple_use_statements.php.inc]: As i was investigating this i thought it has something to do with having multiple use statements. But i think this is not true anymore. I am gonna delete this test fixture here.

File: rules-tests/Renaming/Rector/Name/RenameClassRector/FixtureAutoImportNames/rename_class_with_same_name_but_different_namespace_multiple_use_statements.php.inc (hunk 3 above)
Comment (sabbelasichon) [reply rules-tests/Renaming/Rector/Name/RenameClassRector/FixtureAutoImportNames/rename_class_with_same_name_but_different_namespace_multiple_use_statements.php.inc]: Same as above. Gonna drop it. It is not the root cause.

File: rules/Renaming/NodeManipulator/ClassRenamer.php (hunk 2 above)
Comment (sabbelasichon) [reply rules/Renaming/NodeManipulator/ClassRenamer.php:142]: To be honest, i don´t know why these both methods exists. There seem to be subtle differences.

**Comment by sabbelasichon** [comment]
I have cleaned up the git history by rebasing interactively, so the only difference is wrapping the class instantiation within a class.

**Comment by TomasVotruba** [comment]
Could you add that as a new fixture?
Now we dropped the old one so it's not tested anymore.

Btw, what happens with the original finder method? I would like to see the CI fail to understand the bug. Now it seems it works correctly

**Comment by sabbelasichon** [comment]
Ok. I added a separate fixture and leave the name as it was, so test is failing.

**Comment by sabbelasichon** [comment]
As you can see, the old use statement with the same class name is still there. This would lead to a fatal error in PHP

**Comment by TomasVotruba** [comment]
I see, thank you :+1:

**Comment by sabbelasichon** [comment]
As i said, if i change it to the other method everything is working fine here. Tell me when i should do the change.

**Comment by TomasVotruba** [comment]
> As i said, if i change it to the other method everything is working fine here. Tell me when i should do the change.

Go for it :+1: 

Could you unite those methods to one or improve name of one of them? Now they seems they're doing the same thing, which they don't.

**Comment by sabbelasichon** [comment]
@samsonasik Could you elaborate a little bit more why the new method was introduced here https://github.com/rectorphp/rector/pull/5529 and how we could streamline both of them.

**Comment by samsonasik** [comment]
@sabbelasichon the new `findFirstPreviousOfNode()` was introduced because the `findFirstPrevious()` is doing use existing current statement to locate the callable function, which the the expr will always found in some cases, see:

https://github.com/rectorphp/rector-src/blob/6bd11828e7a5161ecddcd116c6d7bd681d116e22/src/PhpParser/Node/BetterNodeFinder.php#L292-L299

which make it not possible to look up previous node only of existing expr, like in PR https://github.com/rectorphp/rector/pull/5529 done, see:

https://github.com/rectorphp/rector-src/blob/6bd11828e7a5161ecddcd116c6d7bd681d116e22/src/PhpParser/Node/BetterNodeFinder.php#L269-L271

which make it possible to look up previous node only, not a statement/expression. 

I tried to combine them to cover both cases, but it made error in unit tests in other rules so I keep both of them, see:

https://github.com/rectorphp/rector/pull/5543

**Comment by sabbelasichon** [comment]
Could you merge this, but i will come back to it after, to streamline these two methods. But for now i would like to approve and see it in the wild. We are using a lot classes in TYPO3 with the same name in the end but in different location. So it is quite "important" for us to have this fixed.

**Comment by TomasVotruba** [comment]
Sure :+1: thanks for the work and investigation.

**Pull request merged**